### PR TITLE
[homematic] Fix `NullPointerException` in discovery

### LINKED ISSUE #17372: [homematic] NPE on startup
After upgrading from openhab 4.1.x to 4.2.1 homegear binding always crashes on startup of openhab:

```
2024-09-05 06:51:23.184 [INFO ] [org.openhab.core.Activator          ] - Starting openHAB 4.2.1 (Release Build)
2024-09-05 06:51:26.693 [INFO ] [.core.internal.i18n.I18nProviderImpl] - Time zone set to 'Europe/Berlin'.
2024-09-05 06:51:26.744 [INFO ] [.core.internal.i18n.I18nProviderImpl] - Location set to '51.45122561763559,6.283836364746094'.
2024-09-05 06:51:26.750 [INFO ] [.core.internal.i18n.I18nProviderImpl] - Locale set to 'de_DE'.
2024-09-05 06:52:09.126 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'jeelink.items'
2024-09-05 06:52:11.888 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'solaredge.items'
2024-09-05 06:52:12.662 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'homegear.items'
2024-09-05 06:52:13.698 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'default.items'
2024-09-05 06:52:15.853 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'mapdb.persist'
2024-09-05 06:52:20.158 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'jeelink.things'
2024-09-05 06:52:20.509 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'homematic.things'
2024-09-05 06:52:20.696 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'solaredge.things'
2024-09-05 06:52:50.144 [INFO ] [.core.model.lsp.internal.ModelServer] - Started Language Server Protocol (LSP) service on port 5007
2024-09-05 06:52:52.465 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'default.rules'
2024-09-05 06:53:06.831 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'boot.rules'
2024-09-05 06:53:21.331 [INFO ] [e.automation.internal.RuleEngineImpl] - Rule engine started.
2024-09-05 06:53:33.011 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception:
java.lang.NullPointerException: Cannot invoke "org.openhab.binding.homematic.internal.communicator.HomematicGateway.cancelLoadAllDeviceMetadata()" because the return value of "org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler.getGateway()" is null
        at org.openhab.binding.homematic.internal.discovery.HomematicDeviceDiscoveryService.stopScan(HomematicDeviceDiscoveryService.java:112) ~[?:?]
        at org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler.disposeInternal(HomematicBridgeHandler.java:185) ~[?:?]
        at org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler.initializeInternal(HomematicBridgeHandler.java:134) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
        at java.lang.Thread.run(Thread.java:840) [?:?]
2024-09-05 06:55:31.309 [WARN ] [core.thing.internal.ThingManagerImpl] - A thing handler factory claims to support 'homematic:HG-BC-SC-Rd-WM-2' for thing 'homematic:HG-BC-SC-Rd-WM-2:homegear:LEQ0004711' for more than 120s, but the thing type can't be found in the registry. This should be fixed in the binding.
2024-09-05 06:55:31.317 [WARN ] [core.thing.internal.ThingManagerImpl] - Could not normalize configuration for 'homematic:HG-BC-SC-Rd-WM-2:homegear:LEQ0004711' because the thing type was not found in registry.
2024-09-05 06:55:31.410 [WARN ] [core.thing.internal.ThingManagerImpl] - A thing handler factory claims to support 'homematic:HG-BC-SC-Rd-WM-2' for thing 'homematic:HG-BC-SC-Rd-WM-2:homegear:LKF0004711' for more than 120s, but the thing type can't be found in the registry. This should be fixed in the binding.
2024-09-05 06:55:31.413 [WARN ] [core.thing.internal.ThingManagerImpl] - Could not normalize configuration for 'homematic:HG-BC-SC-Rd-WM-2:homegear:LKF0004711' because the thing type was not found in registry.
2024-09-05 06:55:31.453 [WARN ] [core.thing.internal.ThingManagerImpl] - A thing handler factory claims to support 'homematic:HG-BC-SC-Rd-WM-2' for thing 'homematic:HG-BC-SC-Rd-WM-2:homegear:LKF0004712' for more than 120s, but the thing type can't be found in the registry. This should be fixed in the binding.
2024-09-05 06:55:31.461 [WARN ] [core.thing.internal.ThingManagerImpl] - Could not normalize configuration for 'homematic:HG-BC-SC-Rd-WM-2:homegear:LKF0004712' because the thing type was not found in registry.
2024-09-05 06:55:31.502 [WARN ] [core.thing.internal.ThingManagerImpl] - A thing handler factory claims to support 'homematic:HG-BC-SC-Rd-WM-2' for thing 'homematic:HG-BC-SC-Rd-WM-2:homegear:NEQ0004711' for more than 120s, but the thing type can't be found in the registry. This should be fixed in the binding.
...
...
```

The bridge then remains in offline status with message "connection refused". A workaround is to disable the bridge and enable it again. This results in further exceptions but at least the bridge and all connected devices (things) switch to online and work as expected.

```
2024-09-05 07:45:56.237 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.dispose()' on 'org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler@1aeb073': Cannot invoke "org.openhab.binding.homematic.internal.communicator.HomematicGateway.cancelLoadAllDeviceMetadata()" because the return value of "org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler.getGateway()" is null
java.lang.NullPointerException: Cannot invoke "org.openhab.binding.homematic.internal.communicator.HomematicGateway.cancelLoadAllDeviceMetadata()" because the return value of "org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler.getGateway()" is null
        at org.openhab.binding.homematic.internal.discovery.HomematicDeviceDiscoveryService.stopScan(HomematicDeviceDiscoveryService.java:112) ~[?:?]
        at org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler.disposeInternal(HomematicBridgeHandler.java:185) ~[?:?]
        at org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler.dispose(HomematicBridgeHandler.java:177) ~[?:?]
        at jdk.internal.reflect.GeneratedMethodAccessor74.invoke(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:569) ~[?:?]
        at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:147) [bundleFile:?]
        at org.openhab.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
        at java.lang.Thread.run(Thread.java:840) [?:?]
2024-09-05 07:45:56.254 [ERROR] [core.thing.internal.ThingManagerImpl] - Exception occurred while disposing handler of thing 'homematic:bridge:homegear': Cannot invoke "org.openhab.binding.homematic.internal.communicator.HomematicGateway.cancelLoadAllDeviceMetadata()" because the return value of "org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler.getGateway()" is null
java.lang.NullPointerException: Cannot invoke "org.openhab.binding.homematic.internal.communicator.HomematicGateway.cancelLoadAllDeviceMetadata()" because the return value of "org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler.getGateway()" is null
        at org.openhab.binding.homematic.internal.discovery.HomematicDeviceDiscoveryService.stopScan(HomematicDeviceDiscoveryService.java:112) ~[?:?]
        at org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler.disposeInternal(HomematicBridgeHandler.java:185) ~[?:?]
        at org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler.dispose(HomematicBridgeHandler.java:177) ~[?:?]
        at jdk.internal.reflect.GeneratedMethodAccessor74.invoke(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:569) ~[?:?]
        at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:147) [bundleFile:?]
        at org.openhab.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
        at java.lang.Thread.run(Thread.java:840) [?:?]
2024-09-05 07:46:01.634 [INFO ] [ternal.communicator.client.RpcClient] - Interface 'WIRED' on gateway 'homegear' not available, disabling support
2024-09-05 07:46:01.642 [INFO ] [ternal.communicator.client.RpcClient] - Interface 'HMIP' on gateway 'homegear' not available, disabling support
2024-09-05 07:46:01.649 [INFO ] [ternal.communicator.client.RpcClient] - Interface 'CUXD' on gateway 'homegear' not available, disabling support
2024-09-05 07:46:01.662 [INFO ] [ternal.communicator.client.RpcClient] - Interface 'GROUP' on gateway 'homegear' not available, disabling support
2024-09-05 07:46:01.788 [INFO ] [ommunicator.AbstractHomematicGateway] - HmGatewayInfo[id=HOMEGEAR,type=Homegear,firmware=0.7.51-3497,address=VBC8884544,rf=true,wired=false,hmip=false,cuxd=false,group=false]

```

### PR DESCRIPTION
**Pull request by soenkekueper**
Fixed a NPE caused due new check introduced in #17206
fixes #17372

### CODE REVIEW

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java [hunk 1]
```diff
@@ -109,15 +109,17 @@ public int getScanTimeout() {
     public synchronized void stopScan() {
         logger.debug("Stopping Homematic discovery scan");
         disableInstallMode();
-        thingHandler.getGateway().cancelLoadAllDeviceMetadata();
+        if (thingHandler.getGateway() != null) {
+            thingHandler.getGateway().cancelLoadAllDeviceMetadata();
+        }
```
Comment (maniac103) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java:114]: I'd suggest using variables:
```
final HomematicGateway gateway = thingHandler.getGetway();
if (gateway != null) {
    gateway.cancelLoadAllDeviceMetadata();
}
```

Same applies above and below.

Comment (lsiepel) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java:114]: Should not be needed as the block is marked `synchronized`.

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java [hunk 2]
```diff
@@ -74,11 +74,11 @@ protected void startScan() {
      */
     private void enableInstallMode() {
         try {
-            HomematicGateway gateway = thingHandler.getGateway();
             Thing bridge = thingHandler.getThing();
             ThingStatus bridgeStatus = bridge.getStatus();
 
-            if (ThingStatus.ONLINE == bridgeStatus) {
+            if (ThingStatus.ONLINE == bridgeStatus && thingHandler.getGateway() != null) {
+                HomematicGateway gateway = thingHandler.getGateway();
```
Comment (lsiepel) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java:81]: Please revert this and only change :

if (ThingStatus.ONLINE == bridgeStatus && gateway != null) {

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java (hunk 1 above)
Comment (soenkekueper) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java:115]: fixed.

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java (hunk 2 above)
Comment (soenkekueper) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java]: fixed.

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java [hunk 3]
```diff
@@ -109,18 +109,22 @@ public int getScanTimeout() {
...
+        }
         waitForScanFinishing();
         super.stopScan();
     }
 
     private void disableInstallMode() {
         try {
             synchronized (installModeSync) {
-                if (isInInstallMode) {
+                HomematicGateway gateway = thingHandler.getGateway();
```
Comment (maniac103) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java:123]: Minor, but still:

```suggestion
                final HomematicGateway gateway = thingHandler.getGateway();
```

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java [hunk 4]
```diff
@@ -173,10 +177,10 @@ public void waitForScanFinishing() {
      * Starts a thread which loads all Homematic devices connected to the gateway.
      */
     public void loadDevices() {
-        if (loadDevicesFuture == null && thingHandler.getGateway() != null) {
+        HomematicGateway gateway = thingHandler.getGateway();
```
Comment (maniac103) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java:180]:
```suggestion
        final HomematicGateway gateway = thingHandler.getGateway();
```

**Review by maniac103 — APPROVED** [review]
LGTM, just got 2 nitpicks.

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Pull request merged**